### DISCUSSION TITLE
Provider Support for External Ceph

### DISCUSSION THREAD

**Discussion by Nachoprojects** [akash-network/community#632] [Ideas]
Feature request to support external, pre-existing, Ceph clusters. With a potentially common provider architecture to have one existing Ceph cluster, it would make sense to have the Akash provider operator be flexible to support this. This can be done in one of two ways to rapidly support a manual configuration with a final evolution of not requiring the Ceph cluster to be connected in the rook-ceph namespace.

First, is an adjustment to https://github.com/akash-network/provider/blob/e2e52140960762b682c15d5fd67040a7a1527aa7/operator/inventory/ceph.go#L285 to support the ClusterState of "Connected" rather than just "Created" for a Ceph cluster that was manually imported to the rook-ceph namespace with the Rook documented process via https://rook.io/docs/rook/latest-release/CRDs/Cluster/external-cluster/external-cluster/ which required some modification on the auth resources to appropriately configure the client.healthchecker user that rook uses with higher permissions to appropriately create the akash-deployments pool (if desired).

Second, is a full ability to run an external ceph cluster with the standard CSI RBD enablement without the need for rook-ceph entirely. This would require the standard CSI RBD setup via https://docs.ceph.com/en/latest/rbd/rbd-kubernetes/ with the appropriate adjustment to the storage class(es) labeling and naming for beta{1..3}. This would also need to have an external user creation script to import into the k8s secrets to appropriately run the rook-ceph-tools pod inside of akash-services (or other pod with ceph cli) and connect to the external cluster with monitors configured to run "df" commands and so on.

Thanks!

**Comment by Nachoprojects** [comment]
Specific to supporting erasure-coding pools I had to make the following change to ceph.go and then there is a storageclass item.
https://github.com/akash-network/provider/blob/e2e52140960762b682c15d5fd67040a7a1527aa7/operator/inventory/ceph.go#L316 changed to "dataPool" which coincides with the following sc lines. This change is purely for the monitoring aspect as it will pickup the EC metadata pool.

parameters:
  clusterID: rook-ceph
  pool: ec-volumes-metadata
  dataPool: ec62-ssd

These coincide with my underlying previously setup replica 3x pool on NVMe for erasure coding metadata and erasure-coding k6m2 data pools.

**Comment by 31trainman** [comment]
This is rather interesting. What exactly are you trying to accomplish with this change?

**Comment by Nachoprojects** [comment]
@31trainman the desire with the changes in total are to accomplish two things.
1. One is to expand storage flexibility to leverage an external, existing Ceph cluster rather than only rely on the internally Rook Ceph created cluster if an environment already has one. This works today with the ClusterState change in the Akash side as the CSI plugin supports it and this is solely for the inventory to function correctly.
2. The second is to support more than just replicated Ceph storagte pools which are split reporting wise between two different destinations, the metadata pool and then the data pool. This allows for flexibility in storage configurations including using erasure coding which offers sometimes dramatically increased storage utilization efficiency while maintaing same or similar levels of performance.

**Comment by 31trainman** [comment]
Can we discuss this further on akash discord?

My discord is 31trainman
